// vant/es/field/utils.js
import { isObject, isPromise, isFunction, getRootScrollTop, setRootScrollTop } from '../utils';

function isEmptyValue(value) {
  if (Array.isArray(value)) {
    return !value.length;
  }

  if (value === 0) {
    return false;
  }

  return !value;
}

export function runSyncRule(value, rule) {
  if (rule.required && isEmptyValue(value)) {
    return false;
  }

  if (rule.pattern && !rule.pattern.test(String(value))) {
    return false;
  }

  return true;
}
export function runRuleValidator(value, rule) {
  return new Promise(resolve => {
    var returnVal = rule.validator(value, rule);

    if (isPromise(returnVal)) {
      return returnVal.then(resolve);
    }

    resolve(returnVal);
  });
}
export function getRuleMessage(value, rule) {
  var {
    message
  } = rule;

  if (isFunction(message)) {
    return message(value, rule);
  }

  return message || '';
}
export function startComposing(_ref) {
  var {
    target
  } = _ref;
  target.composing = true;
}
export function endComposing(_ref2) {
  var {
    target
  } = _ref2;

  if (target.composing) {
    target.composing = false;
    target.dispatchEvent(new Event('input'));
  }
}
export function resizeTextarea(input, autosize) {
  var scrollTop = getRootScrollTop();
  input.style.height = 'auto';
  var height = input.scrollHeight;

  if (isObject(autosize)) {
    var {
      maxHeight,
      minHeight
    } = autosize;

    if (maxHeight !== undefined) {
      height = Math.min(height, maxHeight);
    }

    if (minHeight !== undefined) {
      height = Math.max(height, minHeight);
    }
  }

  if (height) {
    input.style.height = height + "px"; // https://github.com/youzan/vant/issues/9178

    setRootScrollTop(scrollTop);
  }
}
export function mapInputType(type) {
  // type="number" is weird in iOS, and can't prevent dot in Android
  // so use inputmode to set keyboard in modern browsers
  if (type === 'number') {
    return {
      type: 'text',
      inputmode: 'decimal'
    };
  }

  if (type === 'digit') {
    return {
      type: 'tel',
      inputmode: 'numeric'
    };
  }

  return {
    type
  };
} // get correct length of emoji
// https://github.com/youzan/vant/issues/10032

export function getStringLength(str) {
  return [...str].length;
} // cut string with emoji

export function cutString(str, maxlength) {
  return [...str].slice(0, maxlength).join('');
}
