// element-plus/lib/index.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var defaults = require('./defaults.js');
require('./components/index.js');
require('./directives/index.js');
require('./hooks/index.js');
require('./tokens/index.js');
var popupManager = require('./utils/popup-manager.js');
var makeInstaller = require('./make-installer.js');
var affix = require('./components/affix/src/affix.js');
var index = require('./components/affix/index.js');
var alert = require('./components/alert/src/alert.js');
var index$1 = require('./components/alert/index.js');
var index$2 = require('./components/autocomplete/index.js');
var avatar = require('./components/avatar/src/avatar.js');
var index$3 = require('./components/avatar/index.js');
var backtop = require('./components/backtop/src/backtop.js');
var index$4 = require('./components/backtop/index.js');
var badge = require('./components/badge/src/badge.js');
var index$5 = require('./components/badge/index.js');
var breadcrumb = require('./components/breadcrumb/src/breadcrumb.js');
var breadcrumbItem = require('./components/breadcrumb/src/breadcrumb-item.js');
var index$6 = require('./components/breadcrumb/index.js');
var button = require('./components/button/src/button.js');
var index$7 = require('./components/button/index.js');
var calendar = require('./components/calendar/src/calendar.js');
var index$8 = require('./components/calendar/index.js');
var card = require('./components/card/src/card.js');
var index$9 = require('./components/card/index.js');
var index$a = require('./components/carousel/index.js');
var index$b = require('./components/cascader/index.js');
var types = require('./components/cascader-panel/src/types.js');
var config = require('./components/cascader-panel/src/config.js');
var index$c = require('./components/cascader-panel/index.js');
var index$d = require('./components/check-tag/index.js');
var index$e = require('./components/checkbox/index.js');
var col = require('./components/col/src/col.js');
var index$f = require('./components/col/index.js');
var index$g = require('./components/collapse/index.js');
var index$h = require('./components/collapse-transition/index.js');
var index$i = require('./components/color-picker/index.js');
var configProvider = require('./components/config-provider/src/config-provider.js');
var index$j = require('./components/config-provider/index.js');
var index$k = require('./components/container/index.js');
var index$l = require('./components/date-picker/index.js');
var index$m = require('./components/descriptions/index.js');
var useDialog = require('./components/dialog/src/use-dialog.js');
var dialog = require('./components/dialog/src/dialog.js');
var index$n = require('./components/dialog/index.js');
var divider = require('./components/divider/src/divider.js');
var index$o = require('./components/divider/index.js');
var drawer = require('./components/drawer/src/drawer.js');
var index$p = require('./components/drawer/index.js');
var index$q = require('./components/dropdown/index.js');
var empty = require('./components/empty/src/empty.js');
var index$r = require('./components/empty/index.js');
var index$s = require('./components/form/index.js');
var icon = require('./components/icon/src/icon.js');
var index$t = require('./components/icon/index.js');
var image = require('./components/image/src/image.js');
var index$u = require('./components/image/index.js');
var imageViewer = require('./components/image-viewer/src/image-viewer.js');
var index$v = require('./components/image-viewer/index.js');
var input = require('./components/input/src/input.js');
var index$w = require('./components/input/index.js');
var inputNumber = require('./components/input-number/src/input-number.js');
var index$x = require('./components/input-number/index.js');
var link = require('./components/link/src/link.js');
var index$y = require('./components/link/index.js');
var menu = require('./components/menu/src/menu.js');
var menuItem = require('./components/menu/src/menu-item.js');
var menuItemGroup = require('./components/menu/src/menu-item-group.js');
var subMenu = require('./components/menu/src/sub-menu.js');
var index$z = require('./components/menu/index.js');
var overlay = require('./components/overlay/src/overlay.js');
var index$A = require('./components/overlay/index.js');
var pageHeader = require('./components/page-header/src/page-header.js');
var index$B = require('./components/page-header/index.js');
var pagination = require('./components/pagination/src/pagination.js');
var index$C = require('./components/pagination/index.js');
var popconfirm = require('./components/popconfirm/src/popconfirm.js');
var index$D = require('./components/popconfirm/index.js');
var arrow_vue_vue_type_script_lang = require('./components/popper/src/arrow.vue_vue_type_script_lang.js');
var trigger_vue_vue_type_script_lang = require('./components/popper/src/trigger.vue_vue_type_script_lang.js');
var content_vue_vue_type_script_lang = require('./components/popper/src/content.vue_vue_type_script_lang.js');
var popper = require('./components/popper/src/popper.js');
var tokens = require('./components/popper/src/tokens.js');
var index$E = require('./components/popper/index.js');
var progress = require('./components/progress/src/progress.js');
var index$F = require('./components/progress/index.js');
var radio = require('./components/radio/src/radio.js');
var radioGroup = require('./components/radio/src/radio-group.js');
var radioButton = require('./components/radio/src/radio-button.js');
var index$G = require('./components/radio/index.js');
var rate = require('./components/rate/src/rate.js');
var index$H = require('./components/rate/index.js');
var result = require('./components/result/src/result.js');
var index$I = require('./components/result/index.js');
var row = require('./components/row/src/row.js');
var index$J = require('./components/row/index.js');
var util = require('./components/scrollbar/src/util.js');
var scrollbar = require('./components/scrollbar/src/scrollbar.js');
var bar = require('./components/scrollbar/src/bar.js');
var index$K = require('./components/scrollbar/index.js');
var token = require('./components/select/src/token.js');
var index$L = require('./components/select/index.js');
var token$1 = require('./components/select-v2/src/token.js');
var index$M = require('./components/select-v2/index.js');
var skeleton = require('./components/skeleton/src/skeleton.js');
var skeletonItem = require('./components/skeleton/src/skeleton-item.js');
var index$N = require('./components/skeleton/index.js');
var index$O = require('./components/slider/index.js');
var space = require('./components/space/src/space.js');
var useSpace = require('./components/space/src/use-space.js');
var index$P = require('./components/space/index.js');
var index$Q = require('./components/steps/index.js');
var _switch = require('./components/switch/src/switch.js');
var index$R = require('./components/switch/index.js');
var index$S = require('./components/table/index.js');
var tabs = require('./components/tabs/src/tabs.js');
var tabBar = require('./components/tabs/src/tab-bar.js');
var tabNav = require('./components/tabs/src/tab-nav.js');
var tabPane = require('./components/tabs/src/tab-pane.js');
var index$T = require('./components/tabs/index.js');
var tag = require('./components/tag/src/tag.js');
var index$U = require('./components/tag/index.js');
var dateUtils = require('./components/time-picker/src/common/date-utils.js');
var constant = require('./components/time-picker/src/common/constant.js');
var props = require('./components/time-picker/src/common/props.js');
var picker_vue_vue_type_script_lang = require('./components/time-picker/src/common/picker.vue_vue_type_script_lang.js');
var panelTimePick_vue_vue_type_script_lang = require('./components/time-picker/src/time-picker-com/panel-time-pick.vue_vue_type_script_lang.js');
var index$V = require('./components/time-picker/index.js');
var index$W = require('./components/time-select/index.js');
var index$X = require('./components/timeline/index.js');
var tooltip = require('./components/tooltip/src/tooltip.js');
var index$Y = require('./components/tooltip/index.js');
var constants = require('./utils/constants.js');
var index$Z = require('./components/transfer/index.js');
var index$_ = require('./components/tree/index.js');
var index$$ = require('./components/tree-v2/index.js');
var index$10 = require('./components/upload/index.js');
var fixedSizeList = require('./components/virtual-list/src/components/fixed-size-list.js');
var dynamicSizeList = require('./components/virtual-list/src/components/dynamic-size-list.js');
var fixedSizeGrid = require('./components/virtual-list/src/components/fixed-size-grid.js');
var dynamicSizeGrid = require('./components/virtual-list/src/components/dynamic-size-grid.js');
var props$1 = require('./components/virtual-list/src/props.js');
var index$11 = require('./components/infinite-scroll/index.js');
var index$12 = require('./components/loading/index.js');
var message = require('./components/message/src/message.js');
var index$13 = require('./components/message/index.js');
var index$14 = require('./components/message-box/index.js');
var notification = require('./components/notification/src/notification.js');
var index$15 = require('./components/notification/index.js');
var index$16 = require('./components/popover/index.js');
var index$17 = require('./directives/click-outside/index.js');
var index$18 = require('./directives/repeat-click/index.js');
var index$19 = require('./directives/trap-focus/index.js');
var index$1a = require('./directives/mousewheel/index.js');
var index$1b = require('./directives/resize/index.js');
var index$1c = require('./hooks/use-attrs/index.js');
var index$1d = require('./hooks/use-common-props/index.js');
var index$1e = require('./hooks/use-focus/index.js');
var index$1f = require('./hooks/use-form-item/index.js');
var index$1g = require('./hooks/use-global-config/index.js');
var index$1h = require('./hooks/use-locale/index.js');
var index$1i = require('./hooks/use-lockscreen/index.js');
var index$1j = require('./hooks/use-modal/index.js');
var index$1k = require('./hooks/use-model-toggle/index.js');
var index$1l = require('./hooks/use-prevent-global/index.js');
var index$1m = require('./hooks/use-prop/index.js');
var index$1n = require('./hooks/use-restore-active/index.js');
var index$1o = require('./hooks/use-same-target/index.js');
var index$1p = require('./hooks/use-teleport/index.js');
var index$1q = require('./hooks/use-throttle-render/index.js');
var index$1r = require('./hooks/use-timeout/index.js');
var index$1s = require('./hooks/use-transition-fallthrough/index.js');
var index$1t = require('./hooks/use-id/index.js');
var index$1u = require('./hooks/use-escape-keydown/index.js');
var index$1v = require('./hooks/use-popper-container/index.js');
var index$1w = require('./hooks/use-intermediate-render/index.js');
var index$1x = require('./hooks/use-delayed-toggle/index.js');
var index$1y = require('./hooks/use-forward-ref/index.js');
var form = require('./tokens/form.js');
var button$1 = require('./tokens/button.js');
var breadcrumb$1 = require('./tokens/breadcrumb.js');
var pagination$1 = require('./tokens/pagination.js');
var configProvider$1 = require('./tokens/config-provider.js');
var radio$1 = require('./tokens/radio.js');
var tabs$1 = require('./tokens/tabs.js');
var scrollbar$1 = require('./tokens/scrollbar.js');

const install = defaults["default"].install;
const version = defaults["default"].version;

exports["default"] = defaults["default"];
exports.PopupManager = popupManager.PopupManager;
exports.makeInstaller = makeInstaller.makeInstaller;
exports.affixEmits = affix.affixEmits;
exports.affixProps = affix.affixProps;
exports.ElAffix = index.ElAffix;
exports.alertEmits = alert.alertEmits;
exports.alertProps = alert.alertProps;
exports.ElAlert = index$1.ElAlert;
exports.ElAutocomplete = index$2.ElAutocomplete;
exports.avatarEmits = avatar.avatarEmits;
exports.avatarProps = avatar.avatarProps;
exports.ElAvatar = index$3.ElAvatar;
exports.backtopEmits = backtop.backtopEmits;
exports.backtopProps = backtop.backtopProps;
exports.ElBacktop = index$4.ElBacktop;
exports.badgeProps = badge.badgeProps;
exports.ElBadge = index$5.ElBadge;
exports.breadcrumbProps = breadcrumb.breadcrumbProps;
exports.breadcrumbItemProps = breadcrumbItem.breadcrumbItemProps;
exports.ElBreadcrumb = index$6.ElBreadcrumb;
exports.ElBreadcrumbItem = index$6.ElBreadcrumbItem;
exports.buttonEmits = button.buttonEmits;
exports.buttonNativeType = button.buttonNativeType;
exports.buttonProps = button.buttonProps;
exports.buttonSize = button.buttonSize;
exports.buttonType = button.buttonType;
exports.ElButton = index$7.ElButton;
exports.ElButtonGroup = index$7.ElButtonGroup;
exports.calendarEmits = calendar.calendarEmits;
exports.calendarProps = calendar.calendarProps;
exports.ElCalendar = index$8.ElCalendar;
exports.cardProps = card.cardProps;
exports.ElCard = index$9.ElCard;
exports.ElCarousel = index$a.ElCarousel;
exports.ElCarouselItem = index$a.ElCarouselItem;
exports.ElCascader = index$b.ElCascader;
exports.CASCADER_PANEL_INJECTION_KEY = types.CASCADER_PANEL_INJECTION_KEY;
exports.ExpandTrigger = types.ExpandTrigger;
exports.CommonProps = config.CommonProps;
exports.DefaultProps = config.DefaultProps;
exports.useCascaderConfig = config.useCascaderConfig;
exports.ElCascaderPanel = index$c.ElCascaderPanel;
exports.ElCheckTag = index$d.ElCheckTag;
exports.ElCheckbox = index$e.ElCheckbox;
exports.ElCheckboxButton = index$e.ElCheckboxButton;
exports.ElCheckboxGroup = index$e.ElCheckboxGroup;
exports.colProps = col.colProps;
exports.ElCol = index$f.ElCol;
exports.ElCollapse = index$g.ElCollapse;
exports.ElCollapseItem = index$g.ElCollapseItem;
exports.ElCollapseTransition = index$h.ElCollapseTransition;
exports.ElColorPicker = index$i.ElColorPicker;
exports.configProviderProps = configProvider.configProviderProps;
exports.messageConfig = configProvider.messageConfig;
exports.ElConfigProvider = index$j.ElConfigProvider;
exports.ElAside = index$k.ElAside;
exports.ElContainer = index$k.ElContainer;
exports.ElFooter = index$k.ElFooter;
exports.ElHeader = index$k.ElHeader;
exports.ElMain = index$k.ElMain;
exports.ElDatePicker = index$l.ElDatePicker;
exports.ElDescriptions = index$m.ElDescriptions;
exports.ElDescriptionsItem = index$m.ElDescriptionsItem;
exports.useDialog = useDialog.useDialog;
exports.dialogEmits = dialog.dialogEmits;
exports.dialogProps = dialog.dialogProps;
exports.ElDialog = index$n.ElDialog;
exports.dividerProps = divider.dividerProps;
exports.ElDivider = index$o.ElDivider;
exports.drawerEmits = drawer.drawerEmits;
exports.drawerProps = drawer.drawerProps;
exports.ElDrawer = index$p.ElDrawer;
exports.ElDropdown = index$q.ElDropdown;
exports.ElDropdownItem = index$q.ElDropdownItem;
exports.ElDropdownMenu = index$q.ElDropdownMenu;
exports.emptyProps = empty.emptyProps;
exports.ElEmpty = index$r.ElEmpty;
exports.ElForm = index$s.ElForm;
exports.ElFormItem = index$s.ElFormItem;
exports.iconProps = icon.iconProps;
exports.ElIcon = index$t.ElIcon;
exports.imageEmits = image.imageEmits;
exports.imageProps = image.imageProps;
exports.ElImage = index$u.ElImage;
exports.imageViewerEmits = imageViewer.imageViewerEmits;
exports.imageViewerProps = imageViewer.imageViewerProps;
exports.ElImageViewer = index$v.ElImageViewer;
exports.inputEmits = input.inputEmits;
exports.inputProps = input.inputProps;
exports.ElInput = index$w.ElInput;
exports.inputNumberEmits = inputNumber.inputNumberEmits;
exports.inputNumberProps = inputNumber.inputNumberProps;
exports.ElInputNumber = index$x.ElInputNumber;
exports.linkEmits = link.linkEmits;
exports.linkProps = link.linkProps;
exports.ElLink = index$y.ElLink;
exports.menuEmits = menu.menuEmits;
exports.menuProps = menu.menuProps;
exports.menuItemEmits = menuItem.menuItemEmits;
exports.menuItemProps = menuItem.menuItemProps;
exports.menuItemGroupProps = menuItemGroup.menuItemGroupProps;
exports.subMenuProps = subMenu.subMenuProps;
exports.ElMenu = index$z.ElMenu;
exports.ElMenuItem = index$z.ElMenuItem;
exports.ElMenuItemGroup = index$z.ElMenuItemGroup;
exports.ElSubMenu = index$z.ElSubMenu;
exports.overlayEmits = overlay.overlayEmits;
exports.overlayProps = overlay.overlayProps;
exports.ElOverlay = index$A.ElOverlay;
exports.pageHeaderEmits = pageHeader.pageHeaderEmits;
exports.pageHeaderProps = pageHeader.pageHeaderProps;
exports.ElPageHeader = index$B.ElPageHeader;
exports.paginationEmits = pagination.paginationEmits;
exports.paginationProps = pagination.paginationProps;
exports.ElPagination = index$C.ElPagination;
exports.popconfirmProps = popconfirm.popconfirmProps;
exports.ElPopconfirm = index$D.ElPopconfirm;
exports.ElPopperArrow = arrow_vue_vue_type_script_lang["default"];
exports.ElPopperTrigger = trigger_vue_vue_type_script_lang["default"];
exports.ElPopperContent = content_vue_vue_type_script_lang["default"];
exports.Effect = popper.Effect;
exports.usePopperArrowProps = popper.usePopperArrowProps;
exports.usePopperContentProps = popper.usePopperContentProps;
exports.usePopperCoreConfigProps = popper.usePopperCoreConfigProps;
exports.usePopperProps = popper.usePopperProps;
exports.usePopperTriggerProps = popper.usePopperTriggerProps;
exports.POPPER_CONTENT_INJECTION_KEY = tokens.POPPER_CONTENT_INJECTION_KEY;
exports.POPPER_INJECTION_KEY = tokens.POPPER_INJECTION_KEY;
exports.ElPopper = index$E.ElPopper;
exports.progressProps = progress.progressProps;
exports.ElProgress = index$F.ElProgress;
exports.radioEmits = radio.radioEmits;
exports.radioProps = radio.radioProps;
exports.radioPropsBase = radio.radioPropsBase;
exports.useRadio = radio.useRadio;
exports.radioGroupEmits = radioGroup.radioGroupEmits;
exports.radioGroupProps = radioGroup.radioGroupProps;
exports.radioButtonProps = radioButton.radioButtonProps;
exports.ElRadio = index$G.ElRadio;
exports.ElRadioButton = index$G.ElRadioButton;
exports.ElRadioGroup = index$G.ElRadioGroup;
exports.rateEmits = rate.rateEmits;
exports.rateProps = rate.rateProps;
exports.ElRate = index$H.ElRate;
exports.IconComponentMap = result.IconComponentMap;
exports.IconMap = result.IconMap;
exports.resultProps = result.resultProps;
exports.ElResult = index$I.ElResult;
exports.rowProps = row.rowProps;
exports.ElRow = index$J.ElRow;
exports.BAR_MAP = util.BAR_MAP;
exports.renderThumbStyle = util.renderThumbStyle;
exports.scrollbarEmits = scrollbar.scrollbarEmits;
exports.scrollbarProps = scrollbar.scrollbarProps;
exports.barProps = bar.barProps;
exports.ElScrollbar = index$K.ElScrollbar;
exports.selectGroupKey = token.selectGroupKey;
exports.selectKey = token.selectKey;
exports.ElOption = index$L.ElOption;
exports.ElOptionGroup = index$L.ElOptionGroup;
exports.ElSelect = index$L.ElSelect;
exports.selectV2InjectionKey = token$1.selectV2InjectionKey;
exports.ElSelectV2 = index$M.ElSelectV2;
exports.skeletonProps = skeleton.skeletonProps;
exports.skeletonItemProps = skeletonItem.skeletonItemProps;
exports.ElSkeleton = index$N.ElSkeleton;
exports.ElSkeletonItem = index$N.ElSkeletonItem;
exports.ElSlider = index$O.ElSlider;
exports.spaceProps = space.spaceProps;
exports.useSpace = useSpace.useSpace;
exports.ElSpace = index$P.ElSpace;
exports.ElStep = index$Q.ElStep;
exports.ElSteps = index$Q.ElSteps;
exports.switchEmits = _switch.switchEmits;
exports.switchProps = _switch.switchProps;
exports.ElSwitch = index$R.ElSwitch;
exports.ElTable = index$S.ElTable;
exports.ElTableColumn = index$S.ElTableColumn;
exports.tabsEmits = tabs.tabsEmits;
exports.tabsProps = tabs.tabsProps;
exports.tabBar = tabBar.tabBar;
exports.tabNavProps = tabNav.tabNavProps;
exports.tabPaneProps = tabPane.tabPaneProps;
exports.ElTabPane = index$T.ElTabPane;
exports.ElTabs = index$T.ElTabs;
exports.tagEmits = tag.tagEmits;
exports.tagProps = tag.tagProps;
exports.ElTag = index$U.ElTag;
exports.extractDateFormat = dateUtils.extractDateFormat;
exports.extractTimeFormat = dateUtils.extractTimeFormat;
exports.rangeArr = dateUtils.rangeArr;
exports.DEFAULT_FORMATS_DATE = constant.DEFAULT_FORMATS_DATE;
exports.DEFAULT_FORMATS_DATEPICKER = constant.DEFAULT_FORMATS_DATEPICKER;
exports.DEFAULT_FORMATS_TIME = constant.DEFAULT_FORMATS_TIME;
exports.timePickerDefaultProps = props.timePickerDefaultProps;
exports.CommonPicker = picker_vue_vue_type_script_lang["default"];
exports.TimePickPanel = panelTimePick_vue_vue_type_script_lang["default"];
exports.ElTimePicker = index$V.ElTimePicker;
exports.ElTimeSelect = index$W.ElTimeSelect;
exports.ElTimeline = index$X.ElTimeline;
exports.ElTimelineItem = index$X.ElTimelineItem;
exports.useTooltipContentProps = tooltip.useTooltipContentProps;
exports.useTooltipProps = tooltip.useTooltipProps;
exports.useTooltipTriggerProps = tooltip.useTooltipTriggerProps;
exports.ElTooltip = index$Y.ElTooltip;
exports.CHANGE_EVENT = constants.CHANGE_EVENT;
exports.ElTransfer = index$Z.ElTransfer;
exports.ElTree = index$_.ElTree;
exports.ElTreeV2 = index$$.ElTreeV2;
exports.ElUpload = index$10.ElUpload;
exports.FixedSizeList = fixedSizeList["default"];
exports.DynamicSizeList = dynamicSizeList["default"];
exports.FixedSizeGrid = fixedSizeGrid["default"];
exports.DynamicSizeGrid = dynamicSizeGrid["default"];
exports.virtualizedGridProps = props$1.virtualizedGridProps;
exports.virtualizedListProps = props$1.virtualizedListProps;
exports.virtualizedProps = props$1.virtualizedProps;
exports.virtualizedScrollbarProps = props$1.virtualizedScrollbarProps;
exports.ElInfiniteScroll = index$11.ElInfiniteScroll;
exports.ElLoading = index$12.ElLoading;
exports.ElLoadingDirective = index$12.ElLoadingDirective;
exports.ElLoadingService = index$12.ElLoadingService;
exports.messageEmits = message.messageEmits;
exports.messageProps = message.messageProps;
exports.messageTypes = message.messageTypes;
exports.ElMessage = index$13.ElMessage;
exports.ElMessageBox = index$14.ElMessageBox;
exports.notificationEmits = notification.notificationEmits;
exports.notificationProps = notification.notificationProps;
exports.notificationTypes = notification.notificationTypes;
exports.ElNotification = index$15.ElNotification;
exports.ElPopover = index$16.ElPopover;
exports.ElPopoverDirective = index$16.ElPopoverDirective;
exports.ClickOutside = index$17["default"];
exports.RepeatClick = index$18["default"];
exports.TrapFocus = index$19["default"];
exports.Mousewheel = index$1a["default"];
exports.Resize = index$1b["default"];
exports.useAttrs = index$1c.useAttrs;
exports.useDisabled = index$1d.useDisabled;
exports.useSize = index$1d.useSize;
exports.useSizeProp = index$1d.useSizeProp;
exports.useFocus = index$1e.useFocus;
exports.useFormItem = index$1f.useFormItem;
exports.provideGlobalConfig = index$1g.provideGlobalConfig;
exports.useGlobalConfig = index$1g.useGlobalConfig;
exports.buildLocaleContext = index$1h.buildLocaleContext;
exports.buildTranslator = index$1h.buildTranslator;
exports.translate = index$1h.translate;
exports.useLocale = index$1h.useLocale;
exports.useLockscreen = index$1i.useLockscreen;
exports.useModal = index$1j.useModal;
exports.createModelToggleComposable = index$1k.createModelToggleComposable;
exports.useModelToggle = index$1k.useModelToggle;
exports.useModelToggleEmits = index$1k.useModelToggleEmits;
exports.useModelToggleProps = index$1k.useModelToggleProps;
exports.usePreventGlobal = index$1l.usePreventGlobal;
exports.useProp = index$1m.useProp;
exports.useRestoreActive = index$1n.useRestoreActive;
exports.useSameTarget = index$1o.useSameTarget;
exports.useTeleport = index$1p.useTeleport;
exports.useThrottleRender = index$1q.useThrottleRender;
exports.useTimeout = index$1r.useTimeout;
exports.useTransitionFallthrough = index$1s.useTransitionFallthrough;
exports.useTransitionFallthroughEmits = index$1s.useTransitionFallthroughEmits;
exports.ID_INJECTION_KEY = index$1t.ID_INJECTION_KEY;
exports.useId = index$1t.useId;
exports.useEscapeKeydown = index$1u.useEscapeKeydown;
exports.POPPER_CONTAINER_ID = index$1v.POPPER_CONTAINER_ID;
exports.POPPER_CONTAINER_SELECTOR = index$1v.POPPER_CONTAINER_SELECTOR;
exports.usePopperContainer = index$1v.usePopperContainer;
exports.useDelayedRender = index$1w.useDelayedRender;
exports.useDelayedToggle = index$1x.useDelayedToggle;
exports.useDelayedToggleProps = index$1x.useDelayedToggleProps;
exports.FORWARD_REF_INJECTION_KEY = index$1y.FORWARD_REF_INJECTION_KEY;
exports.useForwardRef = index$1y.useForwardRef;
exports.useForwardRefDirective = index$1y.useForwardRefDirective;
exports.elFormItemKey = form.elFormItemKey;
exports.elFormKey = form.elFormKey;
exports.buttonGroupContextKey = button$1.buttonGroupContextKey;
exports.elBreadcrumbKey = breadcrumb$1.elBreadcrumbKey;
exports.elPaginationKey = pagination$1.elPaginationKey;
exports.configProviderContextKey = configProvider$1.configProviderContextKey;
exports.radioGroupKey = radio$1.radioGroupKey;
exports.tabsRootContextKey = tabs$1.tabsRootContextKey;
exports.scrollbarContextKey = scrollbar$1.scrollbarContextKey;
exports.install = install;
exports.version = version;
