// vant/es/datetime-picker/utils.js
import { extend } from '../utils';
import { pickerSharedProps } from '../picker/Picker';
export var sharedProps = extend({}, pickerSharedProps, {
  filter: Function,
  columnsOrder: Array,
  formatter: {
    type: Function,
    default: (type, value) => value
  }
});
export var pickerInheritKeys = Object.keys(pickerSharedProps);
export function times(n, iteratee) {
  var index = -1;
  var result = Array(n);

  while (++index < n) {
    result[index] = iteratee(index);
  }

  return result;
}
export function getTrueValue(value) {
  if (!value) {
    return 0;
  }

  while (Number.isNaN(parseInt(value, 10))) {
    if (value.length > 1) {
      value = value.slice(1);
    } else {
      return 0;
    }
  }

  return parseInt(value, 10);
}
export var getMonthEndDay = (year, month) => 32 - new Date(year, month - 1, 32).getDate(); // https://github.com/youzan/vant/issues/10013

export var proxyPickerMethods = (picker, callback) => {
  var methods = ['setValues', 'setIndexes', 'setColumnIndex', 'setColumnValue'];
  return new Proxy(picker, {
    get: (target, prop) => {
      if (methods.includes(prop)) {
        return function () {
          target[prop](...arguments);
          callback();
        };
      }

      return target[prop];
    }
  });
};
