// element-plus/es/components/tooltip/src/tooltip.mjs
import { buildProps, definePropType } from '../../../utils/props.mjs';
import '../../popper/index.mjs';
import '../../../hooks/index.mjs';
import { useDelayedToggleProps } from '../../../hooks/use-delayed-toggle/index.mjs';
import { usePopperContentProps, usePopperTriggerProps } from '../../popper/src/popper.mjs';

const triggers = ["hover", "focus", "click", "contextmenu"];
const useTooltipContentProps = {
  ...useDelayedToggleProps,
  ...usePopperContentProps,
  ...buildProps({
    content: {
      type: String,
      default: ""
    },
    rawContent: {
      type: Boolean,
      default: false
    },
    persistent: Boolean,
    ariaLabel: String,
    visible: {
      type: definePropType(Boolean),
      default: null
    },
    transition: {
      type: String,
      default: "el-fade-in-linear"
    },
    teleported: {
      type: Boolean,
      default: true
    }
  })
};
const useTooltipTriggerProps = {
  ...usePopperTriggerProps,
  disabled: Boolean,
  trigger: {
    type: [String, Array],
    default: "hover"
  }
};
const useTooltipProps = buildProps({
  openDelay: {
    type: Number
  },
  visibleArrow: {
    type: Boolean,
    default: void 0
  },
  hideAfter: {
    type: Number,
    default: 200
  },
  showArrow: {
    type: Boolean,
    default: true
  }
});

export { useTooltipContentProps, useTooltipProps, useTooltipTriggerProps };
